### ISSUE TITLE
Long CUDA kernels, long overall run times and `CudnnLegacyConvRunner` related errors with `pjit` on ampere GPUs

### ISSUE THREAD

**Issue by sudhakarsingh27** [google/jax#8654]
Using `pjit` on Ampere GPUs results in longer run times (This issue persists even if I build `jaxlib` from source with arch support explicitly enabled for `8.0` devices). I've created a small program that does distributed matmul ([multi_layer_pjit_matmul.zip](https://github.com/google/jax/files/7584617/multi_layer_pjit_matmul.zip)).  

Following is a profile of the program on Ampere GPUs captured using nsight systems. There are a lot of `cudaFree` calls and I'm not sure if they're necessary. The program takes `10 min` to run, while the actual computation happens at the very end in just a few seconds. 

![Screenshot from 2021-11-22 12-16-58](https://user-images.githubusercontent.com/4879686/142929226-e1a603d7-e9ed-4e19-a8b5-dd84bb8515f7.png)

For comparison,  here's a profile of the same program on Volta GPU:
![Screenshot from 2021-11-22 13-58-27](https://user-images.githubusercontent.com/4879686/142941387-0b6b86af-4381-4c2f-81c3-53c948594ad2.png)
 
Secondly, running another program that uses `flax` and which does spatial parallel convolution using `pjit` on Ampere devices results in the following error. Is this a known issue? Or does anyone have any idea about the underlying cause and any workarounds?

```
2021-11-22 21:32:28.481653: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CudnnLegacyConvRunner cached across multiple StreamExecutors.
2021-11-22 21:32:28.481729: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CudnnLegacyConvRunner cached across multiple StreamExecutors.
2021-11-22 21:32:28.481795: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CudnnLegacyConvRunner cached across multiple StreamExecutors.
2021-11-22 21:32:28.481833: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CudnnLegacyConvRunner cached across multiple StreamExecutors.
2021-11-22 21:32:28.481884: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CudnnLegacyConvRunner cached across multiple StreamExecutors.
2021-11-22 21:32:28.481915: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CudnnLegacyConvRunner cached across multiple StreamExecutors.
2021-11-22 21:32:28.481952: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CudnnLegacyConvRunner cached across multiple StreamExecutors.
2021-11-22 21:32:34.303972: E external/org_tensorflow/tensorflow/compiler/xla/service/collective_ops_utils.h:216] This thread has been waiting for 5000ms for and may be stuck: participant NcclCliqueParticipantData{rendezvous_key=RendezvousKey{run_id=RunId: 72, global_devices=[0,1,2,3,4,5,6,7], num_local_participants=8, collective_op_kind=cross_module, op_id=9}, device_ordinal=1, stream=0x5d7bea0} waiting for all participants to arrive at rendezvous RendezvousKey{run_id=RunId: 72, global_devices=[0,1,2,3,4,5,6,7], num_local_participants=8, collective_op_kind=cross_module, op_id=9}
2021-11-22 21:32:38.481836: F external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2226] Replicated computation launch failed, but not all replicas terminated. Aborting process to work around deadlock. Failure message (there may have been multiple failures, see the error log for all failures): 

CudnnLegacyConvRunner cached across multiple StreamExecutors.
Aborted (core dumped)

```

Version details:
```
jax                              0.2.26
jaxlib                           0.1.74+cuda11.cudnn82
```

```
ii  libcudnn8                   8.2.6.15-1+cuda11.4               amd64        cuDNN runtime libraries
```

**Comment by hawkinsp** [comment]
We think the second of these problems was fixed by https://github.com/tensorflow/tensorflow/commit/eb8025bcaf25d9ddf2ca6a7015a13b179b239c9b, which didn't make it into the last jaxlib release. (You can build a jaxlib from TF / XLA head by editing the `WORKSPACE` file in the JAX root directory; see the comments in that file.)

We don't know what's happening with the first one yet.

**Comment by sudhakarsingh27** [comment]
@hawkinsp I see, I'll check out the fix for the second problem. Thanks!

Just to clarify for the first one, you're already aware of it, right? Let me know if you need any other information from my side.

**Comment by skye** [comment]
Yes, we've reported this to the XLA:GPU team so they can take a look (b/207418070 for internal reference). Thanks for the clear report.

**Comment by ytsheng** [comment]
Hi curious if the second of the problem is included in the latest jaxlib release: 0.1.75?

**Comment by skye** [comment]
Yes, I believe tensorflow/tensorflow@eb8025b is included in jaxlib 0.1.75.

**Comment by hawkinsp** [comment]
I'm guessing this issue is long fixed, because we routinely run distributed matmuls (in the form of transformer models!) on Ampere. Please comment if it's still a problem!
